### PR TITLE
feat: add AI model tier policy

### LINKED ISSUE #216: chore(ai): explicit model policy — router/chat/program/summarize + preview-vs-stable
## Context

\`convex/ai/providers.ts:23-93\` ships one generic primary/fallback pair per provider. Two problems:

1. **No tier-by-task policy**. Current: everything hits Sonnet 4.6 / Gemini 3 Flash / GPT-5.4 primary. World-class split: cheap for routing/classification/summarization, mid for chat, premium for planning/programming.

2. **Preview models in production**. \`primaryModel: "gemini-3-flash-preview"\` — we ship a preview model as the default to our Gemini BYOK majority. One silent API change from Google breaks the whole product.

## Files

- \`convex/ai/providers.ts:23-93\` — current config
- \`convex/ai/coach.ts:192-249\` — \`buildCoachAgents\` / \`buildCoachAgentsForProvider\` consumers
- New (probably): \`docs/ai/model-policy.md\` or an ADR

## Acceptance

- [ ] Write a short policy doc: which model runs which task tier across all 4 providers
- [ ] Tiers: \`router\` (cheapest — Haiku 4.5 / Gemini Flash-lite / GPT-5.4-nano / OpenRouter auto), \`chat\` (mid — Sonnet 4.6 / Gemini Pro / GPT-5.4-mini), \`programming\` (premium — Opus 4.6 / Gemini Pro + high effort / GPT-5.4), \`summarize\` (router tier)
- [ ] Preview-vs-stable rule: \`*-preview\` models MUST NOT be the default primary. Can only be opt-in via user setting or feature flag
- [ ] Move \`gemini-3-flash-preview\` off the default path; replace with stable \`gemini-3-flash\` (or latest stable)
- [ ] Wire \`prepareStep\` to swap models by tier (pairs with #205, #210)
- [ ] Document per-tier cost in the policy doc for future reference

## References

- OpenAI [Compare Models](https://developers.openai.com/api/docs/models/compare)
- Google [Gemini Models](https://ai.google.dev/gemini-api/docs/models)
- Anthropic [pricing tiers](https://platform.claude.com/docs/en/about-claude/pricing)
- [LLM routing: 85% cost reduction case study](https://www.burnwise.io/blog/llm-model-routing-guide)

Related: #190 (intent routing is one consumer of this policy), #205 (effort per step), #210 (prepareStep tool subset)

### PR DESCRIPTION
**Pull request by JeffOtano**
## Summary

- Closes #216 by replacing implicit primary/fallback model defaults with explicit `router`, `chat`, `programming`, and `summarize` tiers.
- Routes chat turns by prompt intent while preserving existing circuit-breaker fallback behavior.
- Removes `gemini-3-flash-preview` from default code paths and documents the model policy.

## Changes

- Added tiered provider policy, preview/default guards, per-model pricing helpers, and tier-aware prompt budgets in `convex/ai/providers.ts`.
- Updated coach agent construction to build tier metadata/agents and use `prepareStep` only for model switching.
- Replaced chat routing with tier routing: trivial `router -> chat`, default `chat -> router`, complex and approval continuation `programming -> chat`.
- Passed separate primary and fallback `prepareStep` callbacks through `streamWithRetry` so fallback attempts use the fallback tier model.
- Updated budget-cap and circuit-breaker cost estimation to reuse centralized per-model pricing.
- Added model-policy documentation in `docs/ai/model-policy.md` and refreshed BYOK UI default model labels.
- Added/updated tests for provider tiers, routing, coach tier metadata, prepareStep selection, budget pricing, circuit-breaker pricing, and retry fallback prepareStep behavior.

## Checklist

- [x] `npx tsc --noEmit` passes
- [x] `npm run lint` passes
- [x] `npm test` passes (existing + new tests)
- [x] New logic has tests (happy path + at least one error/edge case)
- [x] No new or materially expanded file exceeds the 300-line soft cap
- [x] No file exceeds the 400-line hard cap (enforced by CI)
- [x] Functions stay near the 60-line convention
- [x] No new `any` types (enforced by ESLint); `as` casts only at deserialization boundaries
- [ ] Tested in browser (not run; UI change is label/placeholder copy only)
- [x] Commits follow conventional format (`type: description`)
- [x] No unused exports or dead code introduced

## Test Plan

- `npx vitest run convex/ai/providers.test.ts convex/chatProcessing.test.ts convex/ai/coach.test.ts convex/ai/coachModelPolicy.test.ts convex/ai/budgetCap.test.ts convex/ai/circuitBreaker.test.ts convex/ai/runTelemetry.test.ts convex/ai/resilienceCircuitBreaker.test.ts convex/ai/resilienceStreamFailure.test.ts`
- `bash scripts/check-file-size.sh`
- `npm run typecheck`
- `npm run lint`
- `npm test`

## Release Notes

* **New Features**
  * Introduced AI model tiers (router, chat, programming, summarize) for improved provider-specific configurations.
  * Added per-step token usage pricing with separate rates for cached operations.

* **Improvements**
  * Updated default AI models across providers for better performance and cost efficiency.
  * Refactored cost estimation to use provider-specific pricing lookups.
  * Coach agents now intelligently select model tier based on interaction requirements.

* **Documentation**
  * Added comprehensive AI model policy documentation covering tiers, routing, and pricing.

[Review Change Stack](https://app.coderabbit.ai/change-stack/JeffOtano/roni/pull/368)

### CODE REVIEW

File: convex/ai/providers.ts [hunk 1]
```diff
@@ -38,43 +60,228 @@ function normalizeModelId(modelId: string): string {
...
+    },
+  },
+] as const;
+
+export function getModelPricing(provider: ProviderId, modelId: string): ModelPricing | undefined {
+  const normalized = normalizeModelId(modelId);
+  if (provider === "openrouter" && modelId.trim().toLowerCase() === "openrouter/auto") {
+    return MODEL_PRICING[0].pricing;
+  }
+  return MODEL_PRICING.find((entry) => entry.matches.includes(normalized))?.pricing;
```
[the remark this answers is not in the document]
Comment (JeffOtano) [reply convex/ai/providers.ts]: Good catch. I updated `getModelPricing` to match exact model IDs and hyphen-suffixed family variants, then added coverage for dated Claude and OpenRouter/OpenAI IDs. Verified with `npx tsc --noEmit`, `npm run typecheck`, `npm run lint`, targeted pricing tests, and `npm test`. Resolving this thread.

File: convex/ai/providers.ts [hunk 2]
```diff
@@ -38,43 +60,230 @@ function normalizeModelId(modelId: string): string {
...
+      outputUsdPerMillion: 5,
+    },
+  },
+  {
+    matches: ["gemini-2.5-pro"],
+    pricing: {
+      inputUsdPerMillion: 1.25,
+      cacheReadUsdPerMillion: 0.125,
+      cacheWriteUsdPerMillion: 0.125,
+      outputUsdPerMillion: 10,
```
[the remark this answers is not in the document]
Comment (JeffOtano) [reply convex/ai/providers.ts]: Good catch. I changed the Gemini 2.5 Pro enforcement rate to the official high prompt-size tier for Standard pricing, and added a regression assertion for that rate. Verified with `npx tsc --noEmit`, targeted pricing/circuit tests, `npm run typecheck`, `npm run lint`, and `npm test`. Resolving this thread.

File: convex/ai/circuitBreakerCore.ts [hunk 3]
```diff
@@ -244,14 +168,7 @@ export function resolveHalfOpenProbeResult(args: {
...
-    matches.some((candidate) => normalizedModel.startsWith(candidate)),
-  );
-  return match?.rate;
-}
-
-function normalizeModel(model: string): string {
-  return model.trim().toLowerCase().split("/").pop() ?? model.trim().toLowerCase();
+function getModelRate(provider: string, model: string): ModelPricing | undefined {
+  if (isValidProvider(provider)) return getModelPricing(provider, model);
+  return getModelPricing("openrouter", model);
```
[the remark this answers is not in the document]
Comment (JeffOtano) [reply convex/ai/circuitBreakerCore.ts]: Good catch. I updated circuit-breaker cost estimation to fall back to conservative provider pricing when a user override model ID is not in the known pricing table, and added coverage for a preview OpenRouter override. Verified with `npx tsc --noEmit`, targeted pricing/circuit tests, `npm run typecheck`, `npm run lint`, and `npm test`. Resolving this thread.

File: convex/ai/coach.ts [hunk 4]
```diff
@@ -292,48 +309,82 @@ export interface ProviderAgentArgs {
...
+  if (initialTier === "programming" || initialTier === "router") return initialTier;
 
-  const primaryModelName = modelOverride || config.primaryModel;
-  if (!primaryModelName) {
-    throw new Error(`Provider ${provider} requires a model override (no default model)`);
-  }
+  const hasProgrammingToolCall = steps.some((step) =>
+    step.toolCalls?.some((toolCall) => PROGRAMMING_TOOL_NAMES.has(toolCall.toolName)),
+  );
+  return hasProgrammingToolCall ? "programming" : initialTier;
```
[the remark this answers is not in the document]
Comment (JeffOtano) [reply convex/ai/coach.ts:321]: Good catch. I fixed this by making fallback prepareStep routing fixed to the selected fallback tier, while keeping primary chat routes eligible for programming escalation. Added regression coverage that a chat fallback stays chat after a programming tool call, and verified with `npx tsc --noEmit`, focused routing/resilience tests, `npm run typecheck`, `npm run lint`, and `npm test`. Resolving this thread.

File: docs/ai/model-policy.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+## Pricing
+
+Prices are standard USD per 1M tokens. Cached input is the provider's cache-read price when the provider publishes one. Cache-write pricing is included where the provider charges a distinct write price.
+
+| Model                   |                        Input |                  Cached input |                   Cache write |                         Output |
+| ----------------------- | ---------------------------: | ----------------------------: | ----------------------------: | -----------------------------: |
+| `gemini-2.5-flash-lite` |                        $0.10 |                         $0.01 |                         $0.01 |                          $0.40 |
+| `gemini-2.5-flash`      |                        $0.30 |                         $0.03 |                         $0.03 |                          $2.50 |
+| `gemini-2.5-pro`        | $1.25 / $2.50 by prompt size | $0.125 / $0.25 by prompt size | $0.125 / $0.25 by prompt size | $10.00 / $15.00 by prompt size |
```
[the remark this answers is not in the document]
Comment (JeffOtano) [reply docs/ai/model-policy.md]: Good catch. I aligned the policy doc with the runtime estimator so `gemini-2.5-pro` now shows the fixed conservative enforcement rates, and I also addressed the related CodeRabbit nitpicks in `budgetCap.ts`, `chatProcessing.ts`, and `chatProcessing.test.ts`. Verified with `npx tsc --noEmit`, targeted budget/routing tests, `npm run typecheck`, `npm run lint`, and `npm test`. Resolving this thread.
